### PR TITLE
Feature/persisting listening history

### LINKED ISSUE #58: Track listening history locally
Use [`svelte-persistent-writable`](https://github.com/cdellacqua/svelte-persistent-writable). No need to track the entire history, a rolling history of 6 entries is sufficient.

**Acceptance criteria**:
- The "Recently listened" section on the podcast gallery shows up to 6 recently listened podcasts
- Listening history is persisted locally across page reloads

### PR DESCRIPTION
**Pull request by VanishMax**
Closes #58 

What it does and what it does not:
* Implements the play button in EpisodeCard
* Adds a `podcastId` prop to EpisodeCard to track listening history of podcasts, not episodes
* Stores listening history of 6 (at maximum) podcasts
* Doesn't work on Library page. Do we need it there, @illright?

### CODE REVIEW

File: src/lib/widgets/episode/ui/episode-card.svelte [hunk 1]
```diff
@@ -1,17 +1,30 @@
...
+  import { play, pause, audio } from '$lib/entities/audio';
+  import { addToListeningHistory } from '$lib/entities/audio/model/listening-history';
 
   export let episode: Episode;
-  export let playing: boolean;
+  export let podcastId: number | null = null;
+
+  $: playing = $audio?.src === episode.audioUrl && !$audio?.paused && !$audio?.ended;
+
+  const playAudio = () => {
```
Comment (illright) [inline src/lib/widgets/episode/ui/episode-card.svelte:15]: issue: I'm concerned that this duplicates the logic of `features/playback-controls/model/toggle-global-playback`

suggestion: could we edit that one instead and use it here?

File: src/lib/entities/audio/model/listening-history.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Function that adds a podcast id to the beginning of listening history.
+ * It removes any previous occurrences of the same podcast.
+ *
+ * @param podcastId {string} – an URl to add to history
+ */
+export const addToListeningHistory = (podcastId: number) => {
+  listeningHistory.update((history) => {
+    return [podcastId, ...history.filter((id) => id === podcastId)].slice(0, 6);
```
Comment (illright) [inline src/lib/entities/audio/model/listening-history.ts:18]: question: Have you confirmed this works correctly? Looks like it should be a `!==` here

I'd appreciate if you added unit tests to your code as you write it, it's going to be quite a chore to add them later on

**Review by illright — CHANGES_REQUESTED** [review]

**Comment by illright** [comment]
I'm not sure what do you mean by "doesn't work on the library page", or what a library page is, for that matter. We have a "Your Library" page which is supposed to only list favorites, not the history.

**Comment by VanishMax** [comment]
> or what a library page is, for that matter

I meant that I didn't add tracking the history when listening to the podcast on the `/library` page. It has only the liked episodes without connection to podcasts. Therefore, I didn't manage to get the `podcastId` from episodes on `/library` page

**Comment by illright** [comment]
Oh, okay. I think it's good enough for now

**Comment by VanishMax** [comment]
@alkaitagi Everything is fixed, please check

File: src/lib/widgets/episode/ui/episode-card.svelte [hunk 3]
```diff
@@ -4,27 +4,25 @@
...
-      pause();
-    } else {
-      play(episode.audioUrl);
-      if (podcastId) addToListeningHistory(podcastId);
+  const togglePlaying = () => {
+    if ($paused && podcastId) {
+      addToListeningHistory(podcastId);
     }
+
+    toggleGlobalPlayback();
```
Comment (illright) [inline src/lib/widgets/episode/ui/episode-card.svelte:19]: issue: but now it won't know what audio to play.

suggestion: I think we should modify the `toggleGlobalPlayback` function to accept a URL

**Comment by alkaitagi** [comment]
@alkaitagi I think you should merge from `main` because we've greatly simplified the audio API and it will help you.

**Comment by VanishMax** [comment]
@illright you were totally correct with the audio not being played. I fixed that now, and sorry for the delay

File: src/lib/widgets/episode/ui/episode-card.svelte [hunk 4]
```diff
@@ -1,17 +1,28 @@
 <script lang="ts">
+  import type { Episode } from '$lib/shared/api';
   import { PlaybackButton } from '$lib/features/playback-controls';
   import { LikeButton } from '$lib/features/like-episode';
   import { DownloadLink } from '$lib/features/download-episode';
   import { EpisodeCardShell } from '$lib/entities/episode';
-  import type { Episode } from '$lib/shared/api';
+  import { paused } from '$lib/entities/audio';
+  import { addToListeningHistory } from '$lib/entities/audio/model/listening-history';
```
Comment (illright) [inline src/lib/widgets/episode/ui/episode-card.svelte:8]: issue: according to FSD, you cannot go inside a slice from another slice, you must stop at the `index.ts`:
```suggestion
  import { addToListeningHistory } from '$lib/entities/audio';
```

File: src/lib/widgets/episode/ui/episode-card.svelte [hunk 5]
```diff
@@ -1,17 +1,28 @@
...
+  import { toggleGlobalPlayback } from '$lib/features/playback-controls';
```
Comment (illright) [inline src/lib/widgets/episode/ui/episode-card.svelte:9]: nitpick (non-blocking): this import could be merged with another import above from the same module

**Review by illright — CHANGES_REQUESTED** [review]
Minor issue related to FSD isolation principles

**Review by alkaitagi — APPROVED** [review]

**Review by illright — APPROVED** [review]

**Pull request merged**